src/data/scripts/add_context_by_property_type.py

```diff
@@ -116,45 +88,61 @@ def calculate_building_stats():
     with open(property_stats_file_path, "w") as property_stats_file:
         json.dump(stats_by_property_type, property_stats_file)
 
-    return property_stats_file_path
+    property_stats_file_path
 
-def rank_buildings_by_property_type():
+def rank_buildings_by_property_type(
+        building_data: pd.DataFrame, property_types: List[str], grouped_by_prop_type: pd.DataFrame
+    ) -> List[str]:
     """
     Ranks buildings in relation to their property type, then re-exporting the file
 
     Returns the file paths written to
+
+    TODO: Investigate if this should use just the latest year buildings, because we don't want to
+        rank a building that didn't report in the latest year against buildings of a different year
     """
 
     # calculates the statistics for building property types (e.g. average GHG intensity for Hotels)
-    building_stats_path = calculate_building_stats()
-
-    # inputted data
-    building_data = pd.read_csv(input_benchmark_data_csv_path)
-
-    # use pandas to rank each value for each property and store as category+"RankByProperty"
-    for col in building_cols_to_rank:
-        building_data[col +
-                      'RankByPropertyType'] = sorted_by_property_type[col].rank(ascending=False)
+    building_stats_path = calculate_building_stats(property_types, grouped_by_prop_type)
 
     # Mark columns that look like numbers but should be strings as such to prevent decimals showing
-    # up (e.g. zipcode of 60614 or Ward 9)
-    building_data[string_cols] = building_data[string_cols].astype(str)
+    # up (e.g. zipcode of 60614 or Ward 9) and make sure missing data is output as a string
+    building_data[benchmarking_string_cols] = building_data[benchmarking_string_cols].fillna('').astype(str)
 
     # Mark columns as ints that should never show a decimal, e.g. Number of Buildings, Zipcode
-    building_data[int_cols] = building_data[int_cols].astype('Int64')
+    building_data[benchmarking_int_cols] = building_data[benchmarking_int_cols].astype('Int64')
+
+    # use pandas to rank each value for each property and store as category+"RankByProperty"
+    for col in building_cols_to_rank:
+        building_data[col + 'RankByPropertyType'] = grouped_by_prop_type[col].rank(ascending=False)
 
-    building_data.to_csv(input_benchmark_data_csv_path, sep=',', encoding='utf-8', index=False)
+    output_to_csv(building_data, input_benchmark_data_csv_path)
 
     return [ building_stats_path, input_benchmark_data_csv_path ]
 
 
 ###
 ### Main
 ###
-def main():
+def main() -> None:
```

Comment:
Just added some typing and move code into `main()` so i follows best practices, otherwise that code was running on `import`.